### PR TITLE
Add Invariants

### PR DESCRIPTION
**Pull request by shouc**
## Description
Add invariants for core modules. 

Run invariants:
```
forge test --mc Invariant
```

```
Ran 1 test for test/foundry/invariants/LicenseToken.t.sol:LicenseTokenBaseInvariants
[PASS] invariant_notInitializable() (runs: 256, calls: 3840, reverts: 3290)
Suite result: ok. 1 passed; 0 failed; 0 skipped; finished in 615.31ms (596.43ms CPU time)

Ran 1 test for test/foundry/invariants/IPAccount.t.sol:IPAccountPermissionlessNoNftInvariants
[FAIL. Reason: <empty revert data>] invariant_permissionless() (runs: 257, calls: 3840, reverts: 3837)
Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 1.72s (1.71s CPU time)

Ran 3 tests for test/foundry/invariants/LicenseToken.t.sol:LicenseTokenPermissionlessNoTokenInvariants
[PASS] invariant_balanceOfZero() (runs: 256, calls: 3840, reverts: 3262)
[PASS] invariant_impossibleToMint() (runs: 256, calls: 3840, reverts: 3302)
[PASS] invariant_notInitializable() (runs: 256, calls: 3840, reverts: 3285)
Suite result: ok. 3 passed; 0 failed; 0 skipped; finished in 1.78s (1.76s CPU time)

Ran 3 tests for test/foundry/invariants/LicenseToken.t.sol:LicenseTokenPermissionlessOneTokenNotTransferableInvariants
[PASS] invariant_impossibleToMintOrBurn() (runs: 256, calls: 3840, reverts: 3250)
[PASS] invariant_notInitializable() (runs: 256, calls: 3840, reverts: 3295)
[PASS] invariant_notTransferable() (runs: 256, calls: 3840, reverts: 3253)
Suite result: ok. 3 passed; 0 failed; 0 skipped; finished in 1.89s (1.87s CPU time)

Ran 2 tests for test/foundry/invariants/DisputeModule.t.sol:DisputeInvariants
[PASS] invariant_dispute() (runs: 256, calls: 3840, reverts: 0)
[PASS] invariant_dispute_id() (runs: 256, calls: 3840, reverts: 0)
Suite result: ok. 2 passed; 0 failed; 0 skipped; finished in 2.22s (2.20s CPU time)

Ran 2 tests for test/foundry/invariants/IpRoyaltyVault.t.sol:IpRoyaltyVaultInvariant
[PASS] invariant_impossibleChangeOfIpid() (runs: 256, calls: 3840, reverts: 3217)
[PASS] invariant_usdc_balance() (runs: 256, calls: 3840, reverts: 3181)
Suite result: ok. 2 passed; 0 failed; 0 skipped; finished in 2.07s (2.05s CPU time)

Ran 2 tests for test/foundry/invariants/LicenseToken.t.sol:LicenseTokenPermissionlessOneTokenTransferableInvariants
[PASS] invariant_impossibleToMintOrBurn() (runs: 256, calls: 3840, reverts: 3261)
[PASS] invariant_notInitializable() (runs: 256, calls: 3840, reverts: 3271)
Suite result: ok. 2 passed; 0 failed; 0 skipped; finished in 1.20s (1.18s CPU time)

Ran 1 test for test/foundry/invariants/IPAccount.t.sol:IPAccountPermissionlessWithNftInvariants
[PASS] invariant_permissionless() (runs: 256, calls: 3840, reverts: 3834)
Suite result: ok. 1 passed; 0 failed; 0 skipped; finished in 1.83s (1.81s CPU time)

Ran 5 tests for test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryPausedInvariants
[FAIL. Reason: call did not revert as expected] invariant_allRevert() (runs: 257, calls: 3840, reverts: 3218)
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x000000000000000000000000000000000000106d addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3826, reverts: 3249)
[PASS] invariant_nameAndUri() (runs: 256, calls: 3840, reverts: 3250)
[PASS] invariant_registrationDate() (runs: 256, calls: 3840, reverts: 3219)
[PASS] invariant_totalSupply() (runs: 256, calls: 3840, reverts: 3237)
Suite result: FAILED. 3 passed; 2 failed; 0 skipped; finished in 4.85s (4.82s CPU time)

Ran 5 tests for test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryAllRegisteredInvariants
[FAIL. Reason: call did not revert as expected] invariant_allRevert() (runs: 257, calls: 3840, reverts: 3187)
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000000B0b addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[2]
 invariant_checkIpId() (runs: 256, calls: 3832, reverts: 3242)
[PASS] invariant_nameAndUri() (runs: 256, calls: 3840, reverts: 3263)
[PASS] invariant_registrationDate() (runs: 256, calls: 3840, reverts: 3262)
[PASS] invariant_totalSupply() (runs: 256, calls: 3840, reverts: 3256)
Suite result: FAILED. 3 passed; 2 failed; 0 skipped; finished in 6.34s (6.29s CPU time)

Ran 4 tests for test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryInvariants
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000000DFA addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3826, reverts: 2778)
[PASS] invariant_nameAndUri() (runs: 256, calls: 3840, reverts: 2765)
[PASS] invariant_registrationDate() (runs: 256, calls: 3840, reverts: 2796)
[PASS] invariant_totalSupply() (runs: 256, calls: 3840, reverts: 2796)
Suite result: FAILED. 3 passed; 1 failed; 0 skipped; finished in 3.97s (6.71s CPU time)

Ran 11 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWildcardInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3839)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3840)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3839)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3840)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3839)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3840)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3840)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3839)
Suite result: FAILED. 10 passed; 1 failed; 0 skipped; finished in 13.10s (13.04s CPU time)

Ran 4 tests for test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistrySameChainInvariants
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000004539 addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3834, reverts: 2576)
[PASS] invariant_nameAndUri() (runs: 256, calls: 3840, reverts: 2545)
[PASS] invariant_registrationDate() (runs: 256, calls: 3840, reverts: 2543)
[PASS] invariant_totalSupply() (runs: 256, calls: 3840, reverts: 2603)
Suite result: FAILED. 3 passed; 1 failed; 0 skipped; finished in 15.69s (13.27s CPU time)

Ran 11 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleExistTermsInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3832)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3825)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3830)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3830)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3822)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3828)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3823)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3823)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3818)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3828)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3832)
Suite result: FAILED. 10 passed; 1 failed; 0 skipped; finished in 16.03s (12.42s CPU time)

Ran 11 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleBaseInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3828)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3831)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3834)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3833)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3828)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3832)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3835)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3829)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3831)
Suite result: FAILED. 10 passed; 1 failed; 0 skipped; finished in 13.00s (12.50s CPU time)

Ran 12 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWithTermsDiffTemplatesInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3829)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3830)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3834)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3833)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3834)
[PASS] invariant_initialAlwaysAttached() (runs: 256, calls: 3840, reverts: 3830)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3830)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3835)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3829)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3832)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3834)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3830)
Suite result: FAILED. 11 passed; 1 failed; 0 skipped; finished in 7.45s (15.62s CPU time)

Ran 12 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWithTermsInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3840)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3835)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3835)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3833)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3835)
[PASS] invariant_initialAlwaysAttached() (runs: 256, calls: 3840, reverts: 3834)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3839)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3833)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3831)
Suite result: FAILED. 11 passed; 1 failed; 0 skipped; finished in 7.44s (14.75s CPU time)

Ran 11 tests for test/foundry/invariants/LicensingModule.t.sol:LicensingModuleExpensiveInvariant
[PASS] invariant_attachedMustRecorded() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_count() (runs: 256, calls: 3840, reverts: 3837)
[PASS] invariant_derivativeChildRelationship() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_derivativeCounts() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_derivativeParentRelationship() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_licenseConfig() (runs: 256, calls: 3840, reverts: 3838)
[PASS] invariant_licenseTokenSupply() (runs: 256, calls: 3840, reverts: 3836)
[PASS] invariant_nonexistTermId() (runs: 256, calls: 3840, reverts: 3836)
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3836)
[PASS] invariant_onlyAttachableToKnownLicenseTemplates() (runs: 256, calls: 3840, reverts: 3835)
[PASS] invariant_othersIpNotChanged() (runs: 256, calls: 3840, reverts: 3833)
Suite result: FAILED. 10 passed; 1 failed; 0 skipped; finished in 13.40s (13.65s CPU time)

Ran 2 tests for test/foundry/invariants/IPAccount.t.sol:IPAccountPermissionedWithNftInvariants
[PASS] invariant_permissionedStateUpdate() (runs: 256, calls: 3840, reverts: 3112)
[PASS] invariant_permissionedTokenInfo() (runs: 256, calls: 3840, reverts: 3168)
Suite result: ok. 2 passed; 0 failed; 0 skipped; finished in 21.17s (36.29s CPU time)

Ran 19 test suites in 21.18s (135.77s CPU time): 90 tests passed, 13 failed, 0 skipped (103 total tests)

Failing tests:
Encountered 1 failing test in test/foundry/invariants/IPAccount.t.sol:IPAccountPermissionlessNoNftInvariants
[FAIL. Reason: <empty revert data>] invariant_permissionless() (runs: 257, calls: 3840, reverts: 3837)

Encountered 2 failing tests in test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryAllRegisteredInvariants
[FAIL. Reason: call did not revert as expected] invariant_allRevert() (runs: 257, calls: 3840, reverts: 3187)
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000000B0b addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[2]
 invariant_checkIpId() (runs: 256, calls: 3832, reverts: 3242)

Encountered 1 failing test in test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryInvariants
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000000DFA addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3826, reverts: 2778)

Encountered 2 failing tests in test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryPausedInvariants
[FAIL. Reason: call did not revert as expected] invariant_allRevert() (runs: 257, calls: 3840, reverts: 3218)
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x000000000000000000000000000000000000106d addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3826, reverts: 3249)

Encountered 1 failing test in test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistrySameChainInvariants
[FAIL. Reason: <empty revert data>]
        [Sequence]
                sender=0x0000000000000000000000000000000000004539 addr=[test/foundry/invariants/IPAssetRegistry.t.sol:IPAssetRegistryHarness]0x4d04375eCD86c2B81eb0F55B37aA3fAb41CeCBc4 calldata=registerIpAccount(uint8) args=[0]
 invariant_checkIpId() (runs: 256, calls: 3834, reverts: 2576)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleBaseInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3835)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleExistTermsInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3818)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleExpensiveInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3836)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWildcardInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3840)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWithTermsDiffTemplatesInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3832)

Encountered 1 failing test in test/foundry/invariants/LicensingModule.t.sol:LicensingModuleWithTermsInvariant
[FAIL. Reason: <empty revert data>] invariant_nonexistTermIdInTemplate() (runs: 257, calls: 3840, reverts: 3833)
```

## Test Plan 
N/A

## Related Issue
N/A

### CODE REVIEW

File: test/foundry/invariants/IPAccount.t.sol [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+//         harness = new IPAccountReplayHarness();
+//         recorder = harness.recorder();
+//         targetContract(address(harness));
+//     }
+
+//     function invariant_replay() public {
+//         uint256 recorded = recorder.recorded();
+//         assertEq(recorded, 1, "recorded");
+//     }
+// }
```
Comment (kingster-will) [inline test/foundry/invariants/IPAccount.t.sol:281]: If the commented code is not used now, then should be removed.

**Comment by LeoHChen** [comment]
seems like we have quite some failed invariant tests, do we plan to fix them?  @jdubpark @kingster-will

File: test/foundry/invariants/IPAccount.t.sol (hunk 1 above)
Comment (publicqi) [reply test/foundry/invariants/IPAccount.t.sol]: Removed.

**Comment by publicqi** [comment]
> seems like we have quite some failed invariant tests, do we plan to fix them? @jdubpark @kingster-will



> it would be great to add those test into CI once merged and all fixed. @AndyBoWu

There is no failed test now. The failed test was in a previous commit.

**Review by kingster-will — APPROVED** [review]

**Review by jdubpark — APPROVED** [review]

**Pull request merged**